File: pr-app/scripts/ui_auditor.py
import os
import glob
import re
import json

src_dir = "src/pages"
components_dir = "src/components"

screens = glob.glob(os.path.join(src_dir, "**/*.jsx"), recursive=True)

report = {
    "summary": {
        "total_screens": len(screens),
        "compliant": 0,
        "needing_attention": 0
    },
    "findings": []
}

def analyze_screen(filepath):
    with open(filepath, "r", encoding="utf-8") as f:
        content = f.read()

    findings = []
    issues = 0

    # 1. Look for Header / Top Blue Area
    # Heuristic: <header ...> ... </header> or <div className="...app-header... bg-[#1e3a8a]...">
    
    # We will just parse the whole file for components that look like headers
    header_matches = re.finditer(r'<header[^>]*?className=[^>]*?(?:bg-\[\#1e3a8a\]|bg-brand-blue|app-header)[^>]*?>([\s\S]*?)</header>', content)
    
    for match in header_matches:
        header_content = match.group(1)
        # Find all buttons in header
        buttons = re.findall(r'<(?:button|Button)[^>]*>([\s\S]*?)</(?:button|Button)>', header_content)
        for btn in buttons:
            # check what icons or text are inside
            is_back = 'ArrowLeft' in btn or 'navigate(-1)' in btn or 'navigate(' in btn
            is_home = 'Home' in btn or 'navigate(\'/menu\')' in btn
            is_scan = 'Scan' in btn or 'Scanner' in btn
            is_filter = 'Filter' in btn or 'Search' in btn
            is_action = 'Post' in btn or 'Submit' in btn or 'Confirm' in btn or 'Save' in btn
            
            if not (is_back or is_home):
                findings.append({
                    "type": "illegal_top_button",
                    "severity": "high",
                    "detail": "Action or non-navigation button found in top blue header area",
                    "snippet": btn.strip()[:100]
                })
                issues += 1

    # 2. Look for action buttons everywhere
    # Action buttons usually contain specific text
    button_matches = re.finditer(r'<(?:button|Button)([^>]*)>([\s\S]*?)</(?:button|Button)>', content)
    for match in button_matches:
        attrs = match.group(1)
        inner = match.group(2)
        text_content = re.sub(r'<[^>]+>', '', inner).strip()
        
        is_action = re.search(r'\b(Post|Submit|Confirm|Save|Search|Delete|Transfer|Count|Add)\b', text_content, re.IGNORECASE)
        
        # If it's an action button, it must be full width and blue
        if is_action:
            missing_full_width = 'w-full' not in attrs and 'width' not in attrs
            missing_blue_bg = 'bg-brand-blue' not in attrs and 'bg-[#1e3a8a]' not in attrs and 'bg-blue' not in attrs
            missing_white_text = 'text-white' not in attrs and 'color' not in attrs
            
            if missing_full_width or missing_blue_bg or missing_white_text:
                findings.append({
                    "type": "non_compliant_action_button",
                    "severity": "medium",
                    "detail": f"Action button '{text_content}' missing required styling (w-full, primary-blue, text-white)",
                    "snippet": attrs.strip() + " > " + text_content[:50]
                })
                issues += 1

    # 3. Look for Container Padding
    # Usually right after header there's a main or div
    # If we don't see 'p-' or 'px-' or 'pt-' frequently it's bad, but it's hard to statically analyze perfectly.
    # We will flag screens missing standard padding classes in their main wrapper.
    main_matches = re.search(r'<(?:main|div) className="[^"]*(flex-1|h-full)[^"]*">', content)
    if main_matches:
        main_class = main_matches.group(0)
        if 'p-' not in main_class and 'px-' not in main_class and 'pt-' not in main_class:
            findings.append({
                "type": "missing_padding",
                "severity": "low",
                "detail": "Main content wrapper might be missing layout padding tokens (p-4, px-4, pt-4)",
                "snippet": main_class
            })
            issues += 1

    return findings, issues

for screen in screens:
    filename = os.path.basename(screen)
    findings, issue_count = analyze_screen(screen)
    
    if issue_count > 0:
        report["summary"]["needing_attention"] += 1
        report["findings"].append({
            "file": screen,
            "issues": findings
        })
    else:
        report["summary"]["compliant"] += 1

with open("scan_report.json", "w", encoding="utf-8") as f:
    json.dump(report, f, indent=2)

print(f"Scan complete. {report['summary']['total_screens']} screens analyzed.")
print(f"Compliant: {report['summary']['compliant']}, Needing Attention: {report['summary']['needing_attention']}")

File: pr-app/scripts/ui_auditor.mjs
import fs from 'fs';
import path from 'path';
import { parse } from '@babel/parser';
import _traverse from '@babel/traverse';
import { fileURLToPath } from 'url';

const traverse = _traverse.default || _traverse;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const pagesDir = path.join(__dirname, 'src', 'pages');

const getFiles = (dir, files = []) => {
    const list = fs.readdirSync(dir);
    for (const file of list) {
        const fullPath = path.join(dir, file);
        if (fs.statSync(fullPath).isDirectory()) {
            getFiles(fullPath, files);
        } else if (fullPath.endsWith('.jsx')) {
            files.push(fullPath);
        }
    }
    return files;
};

const files = getFiles(pagesDir);

const report = {
    summary: { total_screens: files.length, compliant: 0, needing_attention: 0 },
    findings: []
};

// Top blue area rules: back, home buttons allowed. 
// Action buttons: Post, Submit, Confirm, Save, Delete, Search.
// Full width: class includes w-full.
// Blue bg: bg-brand-blue or bg-[#1e3a8a].
// White text: text-white.

const isActionWord = (str) => /\b(Post|Submit|Confirm|Save|Search|Delete|Transfer|Count|Add)\b/i.test(str);

for (const file of files) {
    const code = fs.readFileSync(file, 'utf-8');
    let ast;
    try {
        ast = parse(code, {
            sourceType: 'module',
            plugins: ['jsx']
        });
    } catch (e) {
        console.error(`Error parsing ${file}:`, e.message);
        continue;
    }

    const issues = [];
    
    traverse(ast, {
        JSXElement(pathNode) {
            const opening = pathNode.node.openingElement;
            const name = opening.name.name;
            
            // 1. Identify Header / Top Blue Area
            let isHeader = false;
            if (name === 'header') isHeader = true;
            
            let classNameAttr = opening.attributes.find(a => a.name && a.name.name === 'className');
            let classStr = '';
            if (classNameAttr && classNameAttr.value && classNameAttr.value.type === 'StringLiteral') {
                classStr = classNameAttr.value.value;
                if (classStr.includes('app-header') || classStr.includes('bg-[#1e3a8a]') || classStr.includes('bg-brand-blue')) {
                    // Check if it's fixed or absolute at the top to confirm it's a header area
                    if (classStr.includes('top-0') || name === 'header' || classStr.includes('app-header')) {
                        isHeader = true;
                    }
                }
            }

            if (isHeader) {
                // Find all buttons inside this header
                pathNode.traverse({
                    JSXElement(innerPath) {
                        const innerName = innerPath.node.openingElement.name.name;
                        if (innerName === 'button' || innerName === 'Button') {
                            // Inspect this button
                            let isAllowed = false;
                            
                            // Check icons (children)
                            innerPath.traverse({
                                JSXOpeningElement(iconPath) {
                                    const iconName = iconPath.node.name.name;
                                    if (iconName === 'ArrowLeft' || iconName === 'Home') {
                                        isAllowed = true;
                                    }
                                }
                            });
                            
                            // Check onClick behavior (e.g. navigate(-1))
                            const onClickAttr = innerPath.node.openingElement.attributes.find(a => a.name && a.name.name === 'onClick');
                            if (onClickAttr && code.substring(onClickAttr.start, onClickAttr.end).includes('navigate(-1)')) {
                                isAllowed = true;
                            }
                            if (onClickAttr && code.substring(onClickAttr.start, onClickAttr.end).includes('navigate(\'/menu\')')) {
                                isAllowed = true;
                            }

                            if (!isAllowed) {
                                issues.push({
                                    type: "illegal_top_button",
                                    severity: "high",
                                    detail: "Non-allowed button (not Back or Home) found in Top Blue Area.",
                                    snippet: `Line ${innerPath.node.loc.start.line}: <${innerName} ...>`
                                });
                            }
                        }
                    }
                });
            }

            // 2. Identify Action Buttons
            if (name === 'button' || name === 'Button') {
                // Extract inner text
                let textContent = '';
                pathNode.traverse({
                    JSXText(textPath) {
                        textContent += textPath.node.value + ' ';
                    }
                });
                
                if (isActionWord(textContent)) {
                    // It's an action button. Check classes.
                    let isFullWidth = classStr.includes('w-full') || classStr.includes('flex-1');
                    let isBlueBg = classStr.includes('bg-brand-blue') || classStr.includes('bg-[#1e3a8a]') || classStr.includes('bg-blue-');
                    let isWhiteText = classStr.includes('text-white');
                    
                    let missing = [];
                    if (!isFullWidth) missing.push('w-full');
                    if (!isBlueBg) missing.push('bg-brand-blue');
                    if (!isWhiteText) missing.push('text-white');
                    
                    if (missing.length > 0) {
                        issues.push({
                            type: "non_compliant_action_button",
                            severity: "medium",
                            detail: `Action button '${textContent.trim()}' is missing required classes: ${missing.join(', ')}`,
                            snippet: `Line ${pathNode.node.loc.start.line}: className="${classStr}"`
                        });
                    }
                    
                    // Check if it's placed inside top blue area (already handled above, but good to cross-check if we want)
                }
            }
            
            // 3. Layout Padding (Heuristic on main visual wrappers)
            if (name === 'main' || (name === 'div' && classStr.includes('flex-1') && classStr.includes('overflow'))) {
                let hasPadding = false;
                if (classStr.match(/\bp-\d+/)) hasPadding = true;
                if (classStr.match(/\bpx-\d+/) && classStr.match(/\bpy-\d+/)) hasPadding = true;
                if (classStr.match(/\bpx-\d+/) && classStr.match(/\bpt-\d+/)) hasPadding = true;
                
                // Allow style objects with padding too (very naive check)
                let styleAttr = opening.attributes.find(a => a.name && a.name.name === 'style');
                if (styleAttr && code.substring(styleAttr.start, styleAttr.end).includes('padding')) {
                    hasPadding = true;
                }

                if (!hasPadding && !classStr.includes('p-')) {
                    issues.push({
                        type: "missing_padding",
                        severity: "low",
                        detail: `Main content wrapper missing padding tokens (e.g. p-4, px-4).`,
                        snippet: `Line ${pathNode.node.loc.start.line}: className="${classStr}"`
                    });
                }
            }
        }
    });
    
    if (issues.length > 0) {
        report.summary.needing_attention++;
        report.findings.push({
            file: path.relative(__dirname, file),
            issues
        });
    } else {
        report.summary.compliant++;
    }
}

fs.writeFileSync('scan_report.json', JSON.stringify(report, null, 2));
console.log(`Scan complete. ${report.summary.total_screens} files analyzed.`);
console.log(`Compliant: ${report.summary.compliant}, Needing Attention: ${report.summary.needing_attention}`);
